lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+// upsert records a new value for the key, or updates an existing value.
+// If the key was previously marked for deletion, that deletion is undone
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
+
+	cs.upserts[key] = value
+}
+
+// delete marks a key for deletion and removes it from upserts and
+// child changesets, if present.
+func (cs *storageDiff) delete(key string) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
+	delete(cs.childChangeSet, key)
+	delete(cs.upserts, key)
+	cs.deletes[key] = true
+}
+
+// deleteChildLimit deletes lexicographical sorted keys from a child trie with
+// a maximum limit, potentially marking the entire child trie for deletion
+// if the limit is exceeded.
+// This method will do not consider keys created during actual block execution
+// https://spec.polkadot.network/chap-host-api#id-version-2-prototype-2
+func (cs *storageDiff) deleteChildLimit(keyToChild string,
+	currentChildKeys []string, limit int) (
+	deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	if limit == -1 {
+		cs.delete(keyToChild)
+		deletedKeys := len(childChanges.upserts) + len(currentChildKeys)
+		return uint32(deletedKeys), true
+	}
+
+	newKeys := maps.Keys(childChanges.upserts)
+	allKeys := append(newKeys, currentChildKeys...)
+	sort.Strings(allKeys)
+
+	for _, k := range allKeys {
+		if limit == 0 {
+			break
+		}
+		childChanges.delete(k)
+		deleted++
+		// Do not consider keys created during actual block execution
+		if !slices.Contains(newKeys, k) {
+			limit--
+		}
+	}
+	cs.childChangeSet[keyToChild] = childChanges
+
+	return deleted, deleted == uint32(len(allKeys))
+}
+
+// clearPrefixInChild clears keys with a specific prefix within a child trie.
+func (cs *storageDiff) clearPrefixInChild(keyToChild string, prefix []byte,
+	childKeys []string, limit int) (deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+	deleted, allDeleted = childChanges.clearPrefix(prefix, childKeys, limit)
+	cs.childChangeSet[keyToChild] = childChanges
+
+	return deleted, allDeleted
+}
+
+// clearPrefix removes all keys matching a specified prefix, within an
+// optional limit. It returns the number of keys deleted and a boolean
+// indicating if all keys with the prefix were removed.
+func (cs *storageDiff) clearPrefix(prefix []byte, trieKeys []string, limit int) (deleted uint32, allDeleted bool) {
```

Comment:
is the `trieKeys` argument necessary?

Reply:
Yes, because we need to know all the keys from the trie, `storageDiff` contains the new keys but `trieKeys` are the current keys